File: tests/api/e2e/tests/foodLog.test.ts
import { TEST_CONFIGURATION } from "../config"
import crypto from 'node:crypto'

describe("Food Log Storage Integration Tests", () => {
    
    test("Happy Path :: Bad Retreives, Creates, Retreives, Edits, Reretrieves, Deletes, Fails Retreive, Redelete succeeds false", async () => {
        const testUserId = crypto.randomUUID()

        const randomId = crypto.randomUUID()

        const badResult = await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs/' + randomId, {
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId
            }
        })

        expect(badResult.status).toBe(404)

        const date = new Date();
        const endDate = new Date(date);
        endDate.setMinutes(endDate.getMinutes() + 5)

        const input = {
            name: "My Food Log",
            labels: ["Some Label", "Some other label"],
            time: {
                start: date.toISOString(),
                end: endDate.toISOString()
            },
            metrics: {
                calories: 500
            }
        }

        const result = await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs' , {
            method: 'POST',
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(input)
        })

        expect(result.status).toBe(200)
        const testItemId = await result.text()
        expect(testItemId.length).toBeGreaterThan(0);

        const storedItemResult = await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs/' + testItemId, {
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId
            }
        })

        expect(storedItemResult.status).toBe(200)
        const storedItem = await storedItemResult.json()
        expect(storedItem).toEqual({ id: testItemId, ...input })

        storedItem.name = "Modified Food Log"
        storedItem.metrics = {
            calories: 400
        }

        const modifiedResult = await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs/' + testItemId, {
            method: 'PUT',
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(storedItem)
        })
        const modified = await modifiedResult.json();
        expect(modified).toEqual(storedItem)

        const reRetreivedResult = await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs/' + testItemId, {
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId
            }
        })

        const reretreived = await reRetreivedResult.json()
        expect(storedItem).toEqual(reretreived)

        const deleteResult = await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs/' + testItemId, {
            method: 'DELETE',
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId
            }
        })

        expect(deleteResult.status).toBe(204);

        const postDeleteReretreive = await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs/' + testItemId, {
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId
            }
        })

        expect(postDeleteReretreive.status).toBe(404)

        const redeleteResult = await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs/' + testItemId, {
            method: 'DELETE',
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId
            }
        })

        expect(redeleteResult.status).toBe(204);
    })

    test("Queries :: can add some logs, and get expected query results", async () => {
        const testUserId = crypto.randomUUID()

        const pastLog = {
            name: "My Food Log",
            labels: [],
            time: {
                start: new Date(1999, 10, 10),
                end: new Date(1999, 10, 11)
            },
            metrics: {
                calories: 500
            }
        }

        const centerLog = {
            name: "My Food Log",
            labels: [],
            time: {
                start: new Date(1999, 10, 15),
                end: new Date(1999, 10, 16)
            },
            metrics: {
                calories: 500
            }
        }

        const futureLog = {
            name: "My Food Log",
            labels: [],
            time: {
                start: new Date(1999, 10, 20),
                end: new Date(1999, 10, 21)
            },
            metrics: {
                calories: 500
            }
        }

        const pastItemId = await(await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs' , {
            method: 'POST',
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(pastLog)
        })).text()

        const centerItemId = await(await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs' , {
            method: 'POST',
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(centerLog)
        })).text()

        const futureItemId = await(await fetch(TEST_CONFIGURATION.API_HOST + '/api/logs' , {
            method: 'POST',
            headers: {
                [TEST_CONFIGURATION.USERID_HEADER]: testUserId,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(futureLog)
        })).text()

        const queryLogs = async (startDate: Date, endDate: Date) => {
            return await(await fetch(TEST_CONFIGURATION.API_HOST + `/api/logs?startDate=${startDate.toISOString()}&endDate=${endDate.toISOString()}`, {
                headers: {
                    [TEST_CONFIGURATION.USERID_HEADER]: testUserId
                }
            })).json()
        }

        const firstTest = await queryLogs(new Date(1999, 10, 15), new Date(1999, 10, 16))
        
        expect(firstTest.length).toBe(1)
        expect(firstTest[0].id).toBe(centerItemId)

        const secondTest = await queryLogs(new Date(1999, 10, 9), new Date(1999, 10, 16))
        
        expect(secondTest.length).toBe(2)
        expect(secondTest.map((x: any) => x.id).sort()).toEqual([pastItemId, centerItemId].sort())
        
        const thirdTest = await queryLogs(new Date(1999, 10, 15), new Date(1999, 10, 30))
        
        expect(thirdTest.length).toBe(2)
        expect(thirdTest.map((x: any) => x.id).sort()).toEqual([centerItemId, futureItemId].sort())

        const wildTest = await queryLogs(new Date(2012, 0, 1), new Date(2012, 11, 31))
        
        expect(wildTest.length).toBe(0)
    })
})

File: tests/e2e/api/foodlog_test.go
package e2e_test

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"testing"
	"time"

	"github.com/google/uuid"
	"github.com/openhealthsuite/diary/internal/config"
	"github.com/openhealthsuite/diary/internal/server"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func Test_CRUD_Logs(xt *testing.T) {
	target_host := "http://localhost:8937"
	useridheader := "x-openfooddiary-userid"

	if os.Getenv("OFD_E2E_TARGET") != "" {
		target_host = os.Getenv("OFD_E2E_TARGET")
	} else {

		config := config.ServerConfiguration{
			Port:                     8937,
			PostgresConnectionString: "",
			SqliteFile:               ":memory:",

			SignoutEndpoint: "/logout",
			UserIdHeader:    useridheader,

			TemplateDirectory: "../../../web/template",
			StaticDirectory:   "../../../web/static",
		}

		srv, err := server.NewServer(&config)
		require.NoError(xt, err)
		quit, err := srv.RunServer()
		defer func() {
			(*quit) <- os.Interrupt
		}()
	}

	xt.Run("Happy Path :: Bad Retreives, Creates, Retreives, Edits, Reretrieves, Deletes, Fails Retreive, Redelete succeeds false", func(t *testing.T) {

		test_user_id := uuid.NewString()

		random_id := uuid.NewString()

		hdrs := map[string]string{
			useridheader: test_user_id,
		}

		hc := http.Client{}

		resp := doReq(&hc, t, "GET", target_host+"/api/logs"+random_id, nil, hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 404, resp.StatusCode)

		date := time.Now()
		endDate := date.Add(5 * time.Minute)

		created_item := map[string]any{
			"name":   "My Food Log",
			"labels": []string{"some label", "other label"},
			"time": map[string]string{
				"start": date.Format(time.RFC3339),
				"end":   endDate.Format(time.RFC3339),
			},
			"metrics": map[string]int64{
				"calories": 500,
			},
		}
		resp = doReq(&hc, t, "POST", target_host+"/api/logs", toJsonBody(t, created_item), hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 200, resp.StatusCode)
		testItemId := toString(t, resp.Body)
		fmt.Println(testItemId)
		assert.Greater(t, len(testItemId), 0)

		resp = doReq(&hc, t, "GET", target_host+"/api/logs/"+testItemId, nil, hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 200, resp.StatusCode)
		storedItem := decode[map[string]any](t, resp.Body)
		assertSubset(t, created_item, storedItem)

		modified_item := map[string]any{
			"name":   "My Updated Food Log",
			"labels": []string{"some other label", "other some label"},
			"time": map[string]string{
				"start": date.Add(-2 * time.Hour).Format(time.RFC3339),
				"end":   endDate.Add(-1 * time.Hour).Format(time.RFC3339),
			},
			"metrics": map[string]int64{
				"calories": 765,
			},
		}

		resp = doReq(&hc, t, "PUT", target_host+"/api/logs/"+testItemId, toJsonBody(t, modified_item), hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 200, resp.StatusCode)

		resp = doReq(&hc, t, "GET", target_host+"/api/logs/"+testItemId, nil, hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 200, resp.StatusCode)
		storedItem = decode[map[string]any](t, resp.Body)
		assertSubset(t, modified_item, storedItem)

		resp = doReq(&hc, t, "DELETE", target_host+"/api/logs/"+testItemId, nil, hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 204, resp.StatusCode)

		resp = doReq(&hc, t, "GET", target_host+"/api/logs/"+testItemId, nil, hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 404, resp.StatusCode)

		resp = doReq(&hc, t, "DELETE", target_host+"/api/logs/"+testItemId, nil, hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 204, resp.StatusCode)

	})

	xt.Run("Queries :: can add some logs, and get expected query results", func(t *testing.T) {

		test_user_id := uuid.NewString()

		hdrs := map[string]string{
			useridheader: test_user_id,
		}

		hc := http.Client{}

		past_log := map[string]any{
			"name":   "Past Log",
			"labels": []string{},
			"time": map[string]any{
				"start": time.Date(1999, time.October, 10, 0, 0, 0, 0, time.UTC).Format(time.RFC3339),
				"end":   time.Date(1999, time.October, 11, 0, 0, 0, 0, time.UTC).Format(time.RFC3339),
			},
			"metrics": map[string]any{
				"calories": int64(500),
			},
		}

		center_log := map[string]any{
			"name":   "Center Log",
			"labels": []string{},
			"time": map[string]any{
				"start": time.Date(1999, time.October, 15, 0, 0, 0, 0, time.UTC).Format(time.RFC3339),
				"end":   time.Date(1999, time.October, 16, 0, 0, 0, 0, time.UTC).Format(time.RFC3339),
			},
			"metrics": map[string]any{
				"calories": int64(500),
			},
		}

		future_log := map[string]any{
			"name":   "Future Log",
			"labels": []string{},
			"time": map[string]any{
				"start": time.Date(1999, time.October, 20, 0, 0, 0, 0, time.UTC).Format(time.RFC3339),
				"end":   time.Date(1999, time.October, 21, 0, 0, 0, 0, time.UTC).Format(time.RFC3339),
			},
			"metrics": map[string]any{
				"calories": int64(500),
			},
		}

		resp := doReq(&hc, t, "POST", target_host+"/api/logs", toJsonBody(t, past_log), hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 200, resp.StatusCode)
		past_log_id := toString(t, resp.Body)

		resp = doReq(&hc, t, "POST", target_host+"/api/logs", toJsonBody(t, center_log), hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 200, resp.StatusCode)
		center_log_id := toString(t, resp.Body)

		resp = doReq(&hc, t, "POST", target_host+"/api/logs", toJsonBody(t, future_log), hdrs)
		defer resp.Body.Close()

		assert.Equal(t, 200, resp.StatusCode)
		future_log_id := toString(t, resp.Body)

		queryLogs := func(start_date, end_date time.Time) []map[string]any {
			resp := doReq(&hc, t, "GET", target_host+"/api/logs?startDate="+start_date.Format(time.RFC3339)+"&endDate="+end_date.Format(time.RFC3339), nil, hdrs)
			defer resp.Body.Close()

			assert.Equal(t, 200, resp.StatusCode)
			return decode[[]map[string]any](t, resp.Body)
		}

		lgs := queryLogs(time.Date(1999, time.October, 14, 0, 0, 0, 0, time.UTC), time.Date(1999, time.October, 16, 0, 0, 0, 0, time.UTC))

		assert.Len(t, lgs, 1)
		assert.Equal(t, center_log_id, lgs[0]["id"].(string))

		lgs = queryLogs(time.Date(1999, time.October, 9, 0, 0, 0, 0, time.UTC), time.Date(1999, time.October, 16, 0, 0, 0, 0, time.UTC))

		assert.Len(t, lgs, 2)
		assert.Equal(t, past_log_id, lgs[0]["id"].(string))
		assert.Equal(t, center_log_id, lgs[1]["id"].(string))

		lgs = queryLogs(time.Date(1999, time.October, 15, 0, 0, 0, 0, time.UTC), time.Date(1999, time.October, 30, 0, 0, 0, 0, time.UTC))

		assert.Len(t, lgs, 2)
		assert.Equal(t, center_log_id, lgs[0]["id"].(string))
		assert.Equal(t, future_log_id, lgs[1]["id"].(string))

		lgs = queryLogs(time.Date(2011, time.October, 15, 0, 0, 0, 0, time.UTC), time.Date(2012, time.October, 30, 0, 0, 0, 0, time.UTC))

		assert.Len(t, lgs, 0)
	})

}

func assertSubset(t *testing.T, sourceItem map[string]any, comparedItem map[string]any) {
	for k, v := range sourceItem {
		storedVal := comparedItem[k]
		var expectedNorm, actualNorm interface{}
		expectedBytes, err := json.Marshal(v)
		require.NoError(t, err)
		actualBytes, err := json.Marshal(storedVal)
		require.NoError(t, err)
		json.Unmarshal(expectedBytes, &expectedNorm)
		json.Unmarshal(actualBytes, &actualNorm)
		assert.Equal(t, expectedNorm, actualNorm)
	}
}

func doReq(hc *http.Client, t *testing.T, method string, url string, body io.Reader, headers map[string]string) *http.Response {
	req, err := http.NewRequest(method, url, body)
	for k, v := range headers {

		req.Header.Set(k, v)
	}
	require.NoError(t, err)
	resp, err := hc.Do(req)
	require.NoError(t, err)
	return resp
}

func toJsonBody(t *testing.T, obj interface{}) io.Reader {
	rs, err := json.Marshal(obj)
	require.NoError(t, err)
	return bytes.NewReader(rs)
}

func toString(t *testing.T, bdy io.Reader) string {
	buf := new(bytes.Buffer)
	_, err := buf.ReadFrom(bdy)
	require.NoError(t, err)
	return buf.String()

}

func decode[E any](t *testing.T, bdy io.Reader) E {
	rs := new(E)
	buf := new(bytes.Buffer)
	_, err := buf.ReadFrom(bdy)
	require.NoError(t, err)
	err = json.Unmarshal(buf.Bytes(), &rs)
	require.NoError(t, err)
	return *rs
}
